pennylane/_qubit_device.py

```diff
@@ -380,18 +380,47 @@ def execute_new(self, circuit, **kwargs):
         # apply all circuit operations
         self.apply(circuit.operations, rotations=circuit.diagonalizing_gates, **kwargs)
 
-        results = self.statistics(circuit.observables)
+        # generate computational basis samples
+        if self.shots is not None or circuit.is_sampled:
+            self._samples = self.generate_samples()
 
-        if len(circuit.measurements) == 1:
-            if circuit.measurements[0].return_type is qml.measurements.State:
-                # State: assumed to only be allowed if it's the only measurement
-                results = self._asarray(results[0], dtype=self.C_DTYPE)
-            else:
-                # All other measurements are real
-                results = self._asarray(results[0], dtype=self.R_DTYPE)
+        # compute the required statistics
+        if not self.analytic and self._shot_vector is not None:
+
+            results = self.shot_vec_statistics(circuit)
 
         else:
-            results = tuple(self._asarray(res, dtype=self.R_DTYPE) for res in results)
+            results = self.statistics_new(circuit.observables)
+
+            ret_types = [m.return_type for m in circuit.measurements]
+            counts_exist = any(ret is qml.measurements.Counts for ret in ret_types)
+            single_measurement = len(circuit.measurements) == 1
+
+            if single_measurement:
+                if circuit.measurements[0].return_type is qml.measurements.State:
+                    # State: assumed to only be allowed if it's the only measurement
+                    results = self._asarray(results[0], dtype=self.C_DTYPE)
```

Comment:
We can maybe move some of the logic directly in the `statistics_new` function (squeezing and asarray)

Reply:
This was an amazing point, made the execute_new logic so much cleaner! :heart_eyes: Thank you